File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode last = reverse(head.next);
        head.next.next = head;
        head.next = null;
        return last;
    }
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) {
            return;
        }
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode rev = reverse(slow);
        ListNode curr = head;
        while (rev.next != null) {
            ListNode tempcurr = curr.next;
            curr.next = rev;

            ListNode temprev = rev.next;
            rev.next = tempcurr;

            curr = tempcurr;
            rev  = temprev;  
        }        
    }
}

// class Solution {
//     public void reorderList(ListNode head) {
//         List<Integer> arr = new ArrayList<>();
//         ListNode temp = head;
//         while (temp != null) {
//             arr.add(temp.val);
//             temp = temp.next;
//         }
//         int n = arr.size();
//         temp = head;
//         int i = 0;
//         while (temp != null && temp.next != null) {
//             temp.val = arr.get(i);
//             temp.next.val = arr.get(n - i - 1);
//             temp = temp.next.next;
//             i++;
//         }
//         if (n % 2 != 0) {
//             temp.val = arr.get(n / 2);
//         }
//     }
// }

File: 0143-reorder-list/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *reverse(ListNode *head) {
        if (head == nullptr || head->next == nullptr) {
            return head;
        }
        ListNode *last = reverse(head->next);
        head->next->next = head;
        head->next = nullptr;
        return last;
    }
    void reorderList(ListNode* head) {
        ListNode *slow = head, *fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode *rev= reverse(slow);
        ListNode *curr = head;
        while (rev->next != nullptr) {
            ListNode *tempcurr = curr->next;
            curr->next = rev;

            ListNode *temprev = rev->next;
            rev->next = tempcurr;

            curr = tempcurr;
            rev = temprev;
        }
    }
};

// class Solution {
// public:
//     void reorderList(ListNode* head) {
//         if (!head || !head->next) {
//             return;
//         }
//         ListNode *temp = head;
//         vector<int> arr;
//         while (temp) {
//             arr.push_back(temp->val);
//             temp = temp->next;
//         }
//         int n = arr.size();
//         temp = head;
//         int i = 0;
//         while (temp && temp->next) {
//             temp->val = arr[i];
//             temp->next->val  = arr[n - i - 1];
//             temp = temp->next->next;
//             i++;
//         }
//         if (n % 2 != 0) {
//             temp->val = arr[(n) / 2];
//         }
//     }
// };